File: src/projectGen.cpp
#include "projectGen.hpp"

void writeMainTS(std::string projName) {
    std::ostringstream codeStream;
    codeStream << "import { NeithServer } from 'https://github.com/anotherSimpleCoder/neith/raw/main/mod.ts';" << std::endl;
    codeStream << "import { routes } from './routes.ts';" << std::endl;
    codeStream << "\n";
    codeStream << "NeithServer" << std::endl;
    codeStream << "\t.fromRoutes(routes)" << std::endl;
    codeStream << "\t.serve(3000);" << std::endl;

    Utils::writeFile(projName + "/main.ts", codeStream.str());
}

void writeRoutes(std::string projName) {
    std::ostringstream codeStream;
    codeStream << "import {Route} from 'https://github.com/anotherSimpleCoder/neith/raw/main/mod.ts';" << std::endl;
    codeStream << "\n";
    codeStream << "export const routes: Route[] = [];" << std::endl;

    Utils::writeFile(projName + "/routes.ts", codeStream.str());
}

void ProjectGen::generate() {
    std::string name;
    std::cout << "Project name: ";
    std::cin >> name;
    std::cout << "\n";
    std::cout << name << std::endl;

    if(!std::filesystem::create_directory(name)) {
        std::cout << "FileError: Couldn't create directory " << name << std::endl;
        return;
    }

    Utils::run("cmd /c cd " + name + " && deno init");
    // Utils::run("cmd /c cd " + name + " && git clone https://github.com/anotherSimpleCoder/neith");
    writeRoutes(name);
    writeMainTS(name);
    Utils::run("cmd /c cd " + name + " && deno cache .");
}

File: src/project_gen.rs
use crate::utils;
use std::fmt::Write;
use std::fs::create_dir;

pub fn generate(proj_name: &str) {
    create_dir(proj_name)
        .expect(format!("FileError: Could not create the directory {}", proj_name).as_str());


    utils::run("cd", &[proj_name, "&&", "deno", "init"]);
    write_routes(proj_name);
    write_main_ts(proj_name);
    write_deno_json(proj_name);
    utils::run("cd", &[proj_name, "&&", "deno", "cache", "."]);
}

fn write_routes(proj_name: &str) {
    let mut code = String::new();
    writeln!(&mut code, "import {{ Route }} from 'https://github.com/anotherSimpleCoder/neith/raw/main/mod.ts';").unwrap();
    writeln!(&mut code, "").unwrap();
    writeln!(&mut code, "export const routes: Route[] = [];").unwrap();

    utils::write_file(format!("{}/routes.ts", proj_name).as_str(), code.as_str());
}

fn write_main_ts(proj_name: &str) {
    let mut code = String::new();
    writeln!(&mut code, "import {{ NeithServer }} from 'https://github.com/anotherSimpleCoder/neith/raw/main/mod.ts';").unwrap();
    writeln!(&mut code, "import {{ routes }} from './routes.ts';").unwrap();
    writeln!(&mut code, "").unwrap();
    writeln!(&mut code, "NeithServer").unwrap();
    writeln!(&mut code, "\t.fromRoutes(routes)").unwrap();
    writeln!(&mut code, "\t.serve(3000);").unwrap();

    utils::write_file(format!("{}/main.ts", proj_name).as_str(), code.as_str());
}

fn write_deno_json(proj_name: &str) {
    let mut code = String::new();
    writeln!(&mut code, "{{").unwrap();
    writeln!(&mut code, "\t\"tasks\": {{").unwrap();
    writeln!(&mut code, "\t\t\"dev\": \"deno run --watch main.ts\"").unwrap();
    writeln!(&mut code, "\t}}").unwrap();
    writeln!(&mut code, "}}").unwrap();


    utils::write_file(format!("{}/deno.json", proj_name).as_str(), code.as_str());
}